src/JsonSchema/Uri/UriResolver.php

```diff
@@ -81,77 +34,124 @@ public function resolve($uri, $baseUri = null)
-        if (isset($components['fragment'])) {
-            $baseComponents['fragment'] = $components['fragment'];
+
+        $uri = $this->parse($baseUri);
+
+        // Replace scheme and host if specified in the new uri
+        if (isset($parts['scheme'])) {
+            $uri['scheme'] = $parts['scheme'];
+        }
+        if (isset($parts['host'])) {
+            $uri['host'] = $parts['host'];
         }
 
-        return $this->generate($baseComponents);
-    }
-    
-    /**
-     * Tries to glue a relative path onto an absolute one
-     *
-     * @param string $relativePath
-     * @param string $basePath
-     * @return string Merged path
-     * @throws UriResolverException
-     */
-    public static function combineRelativePathWithBasePath($relativePath, $basePath)
-    {
-        $relativePath = self::normalizePath($relativePath);
-        if ($relativePath == '') {
-            return $basePath;
+        // Join the base URI path with the new path
+        if (isset($parts['path'])) {
+            if (isset($uri['path'])) {
+                if (substr($uri['path'], -1) !== '/') {
+                    $uri['path'] = rtrim(str_replace(basename($uri['path']), '', $uri['path']), '/') . '/';
+                }
+                $uri['path'] .= ltrim($parts['path'], '/');
+            } else {
+                $uri['path'] = $parts['path'];
+            }
         }
-        if ($relativePath{0} == '/') {
-            return $relativePath;
+
+        // Replace query and fragments
+        if (isset($parts['query'])) {
+            $uri['query'] = $parts['query'];
+        }
+        if (isset($parts['fragment'])) {
+            $uri['fragment'] = $parts['fragment'];
         }
 
-        $basePathSegments = explode('/', $basePath);
+        return $this->build($uri);
+    }
+
+    public function extractLocation($uri)
+    {
+        $parts = $this->parse($uri);
+        unset($parts['fragment']);
+
+        return $this->build($parts);
+    }
+
+    public function extractFragment($uri)
+    {
+        $parts = $this->parse($uri);
 
-        preg_match('|^/?(\.\./(?:\./)*)*|', $relativePath, $match);
-        $numLevelUp = strlen($match[0]) /3 + 1;
-        if ($numLevelUp >= count($basePathSegments)) {
-            throw new UriResolverException(sprintf("Unable to resolve URI '%s' from base '%s'", $relativePath, $basePath));
+        if (isset($parts['fragment'])) {
+            $fragment = $this->build(array('fragment' => $parts['fragment']));
+        } else {
+            $fragment = null;
         }
 
-        $basePathSegments = array_slice($basePathSegments, 0, -$numLevelUp);
-        $path = preg_replace('|^/?(\.\./(\./)*)*|', '', $relativePath);
+        return $fragment;
+    }
 
-        return implode('/', $basePathSegments) . '/' . $path;
+    /**
+     * @param string $uri
+     * @return boolean
+     */
+    public function isValid($uri)
```

Comment:
It was public before, I was trying to avoid making too many public interface changes where possible.